README.md
# iometrics

[![Python](https://github.com/elgalu/iometrics/raw/main/docs/img/badges/language.svg)](https://devdocs.io/python/)

Monitor and log Network and Disks statistics in MegaBytes per second.

## Install

```sh
pip install iometrics
```

## Usage

### Pytorch-lightning integration

```py
from pytorch_lightning import Trainer
from iometrics.pytorch_lightning.callbacks import NetworkAndDiskStatsMonitor

net_disk_stats = NetworkAndDiskStatsMonitor()

trainer = Trainer(callbacks=[net_disk_stats])
```

#### Metrics generated

* **network/recv_MB_per_sec**    – Received MB/s on all relevant network interfaces as a SUM.
* **network/sent_MB_per_sec**    – Sent     MB/s on all relevant network interfaces as a SUM.
* **disk/util%**                 – Disk utilization percentage as the average of all disk devices.
* **disk/read_MB_per_sec**       – Disks read MB/s    as the sum of all disk devices.
* **disk/writ_MB_per_sec**       – Disks written MB/s as the sum of all disk devices.
* **disk/io_read_count_per_sec** – Disks read I/O operations per second    as the sum of all disk devices.
* **disk/io_writ_count_per_sec** – Disks written I/O operations per second as the sum of all disk devices.

#### Screen shot

<img id="png_recv_MB_per_sec" width="450"
 src="https://github.com/elgalu/iometrics/raw/main/docs/img/metrics/network_recv_MB_per_sec.png" />

### Pure-Python implementation (zero dependencies)

#### Quick check

```sh
python -c 'from iometrics.example import usage; usage()'
```

#### Example output

```markdown
|        Network (MBytes/s)       | Disk Util |            Disk MBytes          |           Disk I/O          |
|     Received    |     Sent      |     %     |    MB/s Read    |  MB/s Written |     I/O Read    | I/O Write |
|   val  |   avg  |  val  |  avg  | val | avg |  val   |  avg   |  val  |  avg  |   val  |   avg  | val | avg |
| ------:| ------:| -----:| -----:| ---:| ---:| ------:| ------:| -----:| -----:| ------:| ------:| ---:| ---:|
|    4.6 |    3.5 |   0.1 |   0.1 |  49 |   2 |   52.8 |    1.1 |   0.0 |   0.9 |    211 |      4 |   5 |  18 |
|    4.1 |    3.5 |   0.1 |   0.1 |  61 |   3 |   60.4 |    2.4 |  40.3 |   1.7 |    255 |     10 | 149 |  21 |
```

#### Full code

```py
import time
from iometrics import NetworkMetrics, DiskMetrics
from iometrics.example import DUAL_METRICS_HEADER
net  = NetworkMetrics()
disk = DiskMetrics()
for i in range(100):
    time.sleep(1)
    net.update_stats()
    disk.update_stats()
    if i % 15 == 0:
        print(DUAL_METRICS_HEADER)
    row = (
        f"| {net.mb_recv_ps.val:6.1f} | {net.mb_recv_ps.avg:6.1f} "
        f"| {net.mb_sent_ps.val:5.1f} | {net.mb_sent_ps.avg:5.1f} "
        f"| {int(disk.io_util.val):3d} | {int(disk.io_util.avg):3d} "
        f"| {disk.mb_read.val:6.1f} | {disk.mb_read.avg:6.1f} "
        f"| {disk.mb_writ.val:5.1f} | {disk.mb_writ.avg:5.1f} "
        f"| {int(disk.io_read.val):4d} | {int(disk.io_read.avg):4d} "
        f"| {int(disk.io_writ.val):3d} | {int(disk.io_writ.avg):3d} "
        f"|"
    )
    print(row)
```

## Run in a Docker container

Containers don't have access to the host's network statistics, therefore this workaround is needed.

```sh
# on the host machine (not inside the container)
iometrics replicate proc &
```

after you run above script in the host you should mount the `/host/proc/net/dev` to the container. for example:

```sh
docker run -it -v "/tmp/proc_net_dev:/host/proc/net/dev:ro" <YOURIMAGE>
```

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md)


CHANGELOG.md
# CHANGELOG

## 0.0.8 - [2022-01-20]

### Added

* Add dummy CLI with `iometrics start` and `iometrics replicate proc`  [#cli]

## 0.0.7 - [2021-09-26]

### Added in 0.0.7

* Add TRACK_METRICS_INTERVAL_SECS=10 to track metrics less often  [#TRACK_METRICS_INTERVAL_SECS]

## 0.0.6 - [2021-09-26]

### Added in 0.0.6

* Add "disk/io_wait%" for Disks I/O percentage of time that the CPU is waiting.  [#io_wait]
* Add support for Python 3.7  [#py37]

## 0.0.5 - [2021-09-24]

### Fixed in 0.0.5

* Devices like AWS EBS or USB drives can get dynamically detached so validate key  [#detach]

### Added in 0.0.5

* Add support for Python 3.7  [#py37]

## 0.0.4 - [2021-08-18]

### Fixed in 0.0.4

* Fix bug in PyTorch-Lightning integration  [#PyTorch-Lightning]

## 0.0.3 - [2021-08-17]

### Fixed in 0.0.3

* Fix PyTorch-Lightning integration.  [#PyTorch-Lightning]

## 0.0.2 - [2021-08-15]

### Added on 0.0.2

* Add PyTorch-Lightning support via
`from iometrics.pytorch_lightning.callbacks import NetworkAndDiskStatsMonitor` [#PyTorch-Lightning]

## 0.0.1 - [2021-08-14]

### Added on 0.0.1

* Initial setup release that also tests [tbump](https://github.com/dmerejkowsky/tbump)
  and [towncrier](https://github.com/twisted/towncrier)  [#initial_setup]


iometrics/__init__.py
#!/usr/bin/env python3
"""
## Network and Disk I/O Stats Monitor.

Monitor and log Network and Disks statistics in MegaBytes per second.

### Pytorch-lightning integration

```py
from pytorch_lightning import Trainer
from iometrics.pytorch_lightning.callbacks import NetworkAndDiskStatsMonitor

net_disk_stats = NetworkAndDiskStatsMonitor()

trainer = Trainer(callbacks=[net_disk_stats])
```

### Pure-Python implementation (zero dependencies)

```py
import time
from iometrics import NetworkMetrics, DiskMetrics
from iometrics.example import DUAL_METRICS_HEADER
net  = NetworkMetrics()
disk = DiskMetrics()
for i in range(100):
    time.sleep(1)
    net.update_stats()
    disk.update_stats()
    if i % 15 == 0:
        print(DUAL_METRICS_HEADER)
    row = (
        f"| {net.mb_recv_ps.val:6.1f} | {net.mb_recv_ps.avg:6.1f} "
        f"| {net.mb_sent_ps.val:5.1f} | {net.mb_sent_ps.avg:5.1f} "
        f"| {int(disk.io_util.val):3d} | {int(disk.io_util.avg):3d} "
        f"| {disk.mb_read.val:6.1f} | {disk.mb_read.avg:6.1f} "
        f"| {disk.mb_writ.val:5.1f} | {disk.mb_writ.avg:5.1f} "
        f"| {int(disk.io_read.val):4d} | {int(disk.io_read.avg):4d} "
        f"| {int(disk.io_writ.val):3d} | {int(disk.io_writ.avg):3d} "
        f"|"
    )
    print(row)
```

#### Example output

```md
|        Network (MBytes/s)       | Disk Util |            Disk MBytes          |           Disk I/O          |
|     Received    |     Sent      |     %     |    MB/s Read    |  MB/s Written |     I/O Read    | I/O Write |
|   val  |   avg  |  val  |  avg  | val | avg |  val   |  avg   |  val  |  avg  |   val  |   avg  | val | avg |
| ------:| ------:| -----:| -----:| ---:| ---:| ------:| ------:| -----:| -----:| ------:| ------:| ---:| ---:|
|    4.6 |    3.5 |   0.1 |   0.1 |  49 |   2 |   52.8 |    1.1 |   0.0 |   0.9 |    211 |      4 |   5 |  18 |
|    4.1 |    3.5 |   0.1 |   0.1 |  61 |   3 |   60.4 |    2.4 |  40.3 |   1.7 |    255 |     10 | 149 |  21 |
```

:copyright: (c) 2021 by Leo Gallucci.
:license: Apache 2.0, see LICENSE for more details.
"""
# Do not change the version here but rather `tbump "0.0.5" --only-patch`
__version__ = "0.0.8"

from iometrics.network import NetworkMetrics
from iometrics.disk import DiskMetrics

# `__all__` is left here for documentation purposes and as a
# reference to which interfaces are meant to be imported.
__all__ = [
    "__version__",
    "NetworkMetrics",
    "DiskMetrics",
]
